File: leet_code_solutions.js/20.js
// Solved.
// Runtime: better than 90.09% of JS solutions
// Memory Usage: better than 11.24% of JS solutions

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  if (s.length === 1) {
    return false;
  }
  var stack = [];
  s = s.split("");

  for (let bracket of s) {
    if (bracket === "{" || bracket === "[" || bracket === "(") {
      stack.push(bracket);
    } else {
      var open = stack.pop();
      if (
        (bracket === "}" && open === "{") ||
        (bracket === ")" && open === "(") ||
        (bracket === "]" && open === "[")
      ) {
        continue;
      } else {
        return false;
      }
    }
  }
  if (stack.length !== 0) {
    return false;
  }
  return true;
};

console.log(
  isValid("[(((((())))))][][][][{{{{{{}}}}}}][(()()()({}{}{}{}{}{})]")
);
console.log(isValid("(("));
console.log(isValid("()"));

File: leet_code_solutions.py/20.py
## Runtime: 24 ms, faster than 95.82% of Python3 online submissions for Valid Parentheses.
## Memory Usage: 14.3 MB, less than 62.87% of Python3 online submissions for Valid Parentheses.

class Solution:
    def isValid(self, s: str) -> bool:
        need = []
        
        for i in range(len(s)):
            if s[i]==')' or s[i]=='}' or s[i]==']':
                if len(need)==0:
                    return False
                
                if s[i]==need[-1]:
                    need.pop()
                else:
                    return False
                
            if s[i]=='(' or s[i]=='[' or s[i]=='{':
                
                if ord(s[i])==40:
                    need.append(chr(ord(s[i])+1))
                else:
                    need.append(chr(ord(s[i])+2))
                
        if len(need)==0:
            return True
        return False
